# Data Cleaning Project ( Ebay car sales)

In [1]:
import numpy as np
import pandas as pd


In [2]:
#reading the csv file
data=pd.read_csv('autos.csv')

In [3]:
#checking to see how the data looks like
data.head()

datecrawled                            name  seller offertype  \
0  2016-03-24T11:52:17                      Golf_3_1.6  privat   Angebot   
1  2016-03-24T10:58:45            A5_Sportback_2.7_Tdi  privat   Angebot   
2  2016-03-14T12:52:21  Jeep_Grand_Cherokee_"Overland"  privat   Angebot   
3  2016-03-17T16:54:04              GOLF_4_1_4__3TÜRER  privat   Angebot   
4  2016-03-31T17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat   Angebot   

   price abtest vehicletype  yearofregistration    gearbox  powerps  model  \
0    480   test         NaN                1993    manuell        0   golf   
1  18300   test       coupe                2011    manuell      190    NaN   
2   9800   test         suv                2004  automatik      163  grand   
3   1500   test  kleinwagen                2001    manuell       75   golf   
4   3600   test  kleinwagen                2008    manuell       69  fabia   

   kilometer  monthofregistration fueltype       brand notrepaireddamage  \
0     150000                    0   benzin  volkswagen               NaN   
1     125000                    5   diesel        audi                ja   
2     125000                    8   diesel        jeep               NaN   
3     150000                    6   benzin  volkswagen              nein   
4      90000                    7   diesel       skoda              nein   

           datecreated  nrofpictures  postalcode             lastseen  
0  2016-03-24T00:00:00         False       70435  2016-04-07T03:16:57  
1  2016-03-24T00:00:00         False       66954  2016-04-07T01:46:50  
2  2016-03-14T00:00:00         False       90480  2016-04-05T12:47:46  
3  2016-03-17T00:00:00         False       91074  2016-03-17T17:40:17  
4  2016-03-31T00:00:00         False       60437  2016-04-06T10:17:21

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 20 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   datecrawled          371528 non-null  object
 1   name                 371528 non-null  object
 2   seller               371528 non-null  object
 3   offertype            371528 non-null  object
 4   price                371528 non-null  int64 
 5   abtest               371528 non-null  object
 6   vehicletype          333659 non-null  object
 7   yearofregistration   371528 non-null  int64 
 8   gearbox              351319 non-null  object
 9   powerps              371528 non-null  int64 
 10  model                351044 non-null  object
 11  kilometer            371528 non-null  int64 
 12  monthofregistration  371528 non-null  int64 
 13  fueltype             338142 non-null  object
 14  brand                371528 non-null  object
 15  notrepaireddamage    299468 non-nu

### checking columns to see if we need to change anything in order for them to become more readable

In [5]:
#displaying columns
data.columns


Index(['datecrawled', 'name', 'seller', 'offertype', 'price', 'abtest',
       'vehicletype', 'yearofregistration', 'gearbox', 'powerps', 'model',
       'kilometer', 'monthofregistration', 'fueltype', 'brand',
       'notrepaireddamage', 'datecreated', 'nrofpictures', 'postalcode',
       'lastseen'],
      dtype='object')

In [6]:
#making columns more readable
data.columns=['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'year_of_registration', 'gearbox', 'power_ps', 'model',
       'kilometer', 'month_of_registration', 'fuel_type', 'brand',
       'not_repaired_damage', 'date_created', 'nr_of_pictures', 'postal_code',
       'last_seen']

In [7]:
data.head()


date_crawled                            name  seller offer_type  \
0  2016-03-24T11:52:17                      Golf_3_1.6  privat    Angebot   
1  2016-03-24T10:58:45            A5_Sportback_2.7_Tdi  privat    Angebot   
2  2016-03-14T12:52:21  Jeep_Grand_Cherokee_"Overland"  privat    Angebot   
3  2016-03-17T16:54:04              GOLF_4_1_4__3TÜRER  privat    Angebot   
4  2016-03-31T17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat    Angebot   

   price ab_test vehicle_type  year_of_registration    gearbox  power_ps  \
0    480    test          NaN                  1993    manuell         0   
1  18300    test        coupe                  2011    manuell       190   
2   9800    test          suv                  2004  automatik       163   
3   1500    test   kleinwagen                  2001    manuell        75   
4   3600    test   kleinwagen                  2008    manuell        69   

   model  kilometer  month_of_registration fuel_type       brand  \
0   golf     150000                      0    benzin  volkswagen   
1    NaN     125000                      5    diesel        audi   
2  grand     125000                      8    diesel        jeep   
3   golf     150000                      6    benzin  volkswagen   
4  fabia      90000                      7    diesel       skoda   

  not_repaired_damage         date_created  nr_of_pictures  postal_code  \
0                 NaN  2016-03-24T00:00:00           False        70435   
1                  ja  2016-03-24T00:00:00           False        66954   
2                 NaN  2016-03-14T00:00:00           False        90480   
3                nein  2016-03-17T00:00:00           False        91074   
4                nein  2016-03-31T00:00:00           False        60437   

             last_seen  
0  2016-04-07T03:16:57  
1  2016-04-07T01:46:50  
2  2016-04-05T12:47:46  
3  2016-03-17T17:40:17  
4  2016-04-06T10:17:21

In [8]:
data.describe(include='all')

date_crawled         name  seller offer_type         price  \
count                371528       371528  371528     371528  3.715280e+05   
unique               280500       233531       2          2           NaN   
top     2016-03-24T14:49:47  Ford_Fiesta  privat    Angebot           NaN   
freq                      7          657  371525     371516           NaN   
mean                    NaN          NaN     NaN        NaN  1.729514e+04   
std                     NaN          NaN     NaN        NaN  3.587954e+06   
min                     NaN          NaN     NaN        NaN  0.000000e+00   
25%                     NaN          NaN     NaN        NaN  1.150000e+03   
50%                     NaN          NaN     NaN        NaN  2.950000e+03   
75%                     NaN          NaN     NaN        NaN  7.200000e+03   
max                     NaN          NaN     NaN        NaN  2.147484e+09   

       ab_test vehicle_type  year_of_registration  gearbox       power_ps  \
count   371528       333659         371528.000000   351319  371528.000000   
unique       2            8                   NaN        2            NaN   
top       test    limousine                   NaN  manuell            NaN   
freq    192585        95894                   NaN   274214            NaN   
mean       NaN          NaN           2004.577997      NaN     115.549477   
std        NaN          NaN             92.866598      NaN     192.139578   
min        NaN          NaN           1000.000000      NaN       0.000000   
25%        NaN          NaN           1999.000000      NaN      70.000000   
50%        NaN          NaN           2003.000000      NaN     105.000000   
75%        NaN          NaN           2008.000000      NaN     150.000000   
max        NaN          NaN           9999.000000      NaN   20000.000000   

         model      kilometer  month_of_registration fuel_type       brand  \
count   351044  371528.000000          371528.000000    338142      371528   
unique     251            NaN                    NaN         7          40   
top       golf            NaN                    NaN    benzin  volkswagen   
freq     30070            NaN                    NaN    223857       79640   
mean       NaN  125618.688228               5.734445       NaN         NaN   
std        NaN   40112.337051               3.712412       NaN         NaN   
min        NaN    5000.000000               0.000000       NaN         NaN   
25%        NaN  125000.000000               3.000000       NaN         NaN   
50%        NaN  150000.000000               6.000000       NaN         NaN   
75%        NaN  150000.000000               9.000000       NaN         NaN   
max        NaN  150000.000000              12.000000       NaN         NaN   

       not_repaired_damage         date_created nr_of_pictures   postal_code  \
count               299468               371528         371528  371528.00000   
unique                   2                  114              1           NaN   
top                   nein  2016-04-03T00:00:00          False           NaN   
freq                263182                14450         371528           NaN   
mean                   NaN                  NaN            NaN   50820.66764   
std                    NaN                  NaN            NaN   25799.08247   
min                    NaN                  NaN            NaN    1067.00000   
25%                    NaN                  NaN            NaN   30459.00000   
50%                    NaN                  NaN            NaN   49610.00000   
75%                    NaN                  NaN            NaN   71546.00000   
max                    NaN                  NaN            NaN   99998.00000   

                  last_seen  
count                371528  
unique               182806  
top     2016-04-06T13:45:54  
freq                     17  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN

There are some columns where almost all values are the same, like 
- seller
- offer type
- nr of pictures

These are also columns that are not so important for our analysis so we might drop them 

In [9]:
print(data['nr_of_pictures'].head())
print('\n')
print(data['seller'].head())
print('\n')
print(data['offer_type'].head())


0    False
1    False
2    False
3    False
4    False
Name: nr_of_pictures, dtype: bool


0    privat
1    privat
2    privat
3    privat
4    privat
Name: seller, dtype: object


0    Angebot
1    Angebot
2    Angebot
3    Angebot
4    Angebot
Name: offer_type, dtype: object


As we can see the values in those three columns, serve no purpose for our analysis and they are all the same. As a result we are going to drop all three of them

In [10]:
#dropping the columns
data.drop(['nr_of_pictures','seller','offer_type'], axis=1,inplace=True)

In [11]:
data.describe(include='all')

date_crawled         name         price ab_test vehicle_type  \
count                371528       371528  3.715280e+05  371528       333659   
unique               280500       233531           NaN       2            8   
top     2016-03-24T14:49:47  Ford_Fiesta           NaN    test    limousine   
freq                      7          657           NaN  192585        95894   
mean                    NaN          NaN  1.729514e+04     NaN          NaN   
std                     NaN          NaN  3.587954e+06     NaN          NaN   
min                     NaN          NaN  0.000000e+00     NaN          NaN   
25%                     NaN          NaN  1.150000e+03     NaN          NaN   
50%                     NaN          NaN  2.950000e+03     NaN          NaN   
75%                     NaN          NaN  7.200000e+03     NaN          NaN   
max                     NaN          NaN  2.147484e+09     NaN          NaN   

        year_of_registration  gearbox       power_ps   model      kilometer  \
count          371528.000000   351319  371528.000000  351044  371528.000000   
unique                   NaN        2            NaN     251            NaN   
top                      NaN  manuell            NaN    golf            NaN   
freq                     NaN   274214            NaN   30070            NaN   
mean             2004.577997      NaN     115.549477     NaN  125618.688228   
std                92.866598      NaN     192.139578     NaN   40112.337051   
min              1000.000000      NaN       0.000000     NaN    5000.000000   
25%              1999.000000      NaN      70.000000     NaN  125000.000000   
50%              2003.000000      NaN     105.000000     NaN  150000.000000   
75%              2008.000000      NaN     150.000000     NaN  150000.000000   
max              9999.000000      NaN   20000.000000     NaN  150000.000000   

        month_of_registration fuel_type       brand not_repaired_damage  \
count           371528.000000    338142      371528              299468   
unique                    NaN         7          40                   2   
top                       NaN    benzin  volkswagen                nein   
freq                      NaN    223857       79640              263182   
mean                 5.734445       NaN         NaN                 NaN   
std                  3.712412       NaN         NaN                 NaN   
min                  0.000000       NaN         NaN                 NaN   
25%                  3.000000       NaN         NaN                 NaN   
50%                  6.000000       NaN         NaN                 NaN   
75%                  9.000000       NaN         NaN                 NaN   
max                 12.000000       NaN         NaN                 NaN   

               date_created   postal_code            last_seen  
count                371528  371528.00000               371528  
unique                  114           NaN               182806  
top     2016-04-03T00:00:00           NaN  2016-04-06T13:45:54  
freq                  14450           NaN                   17  
mean                    NaN   50820.66764                  NaN  
std                     NaN   25799.08247                  NaN  
min                     NaN    1067.00000                  NaN  
25%                     NaN   30459.00000                  NaN  
50%                     NaN   49610.00000                  NaN  
75%                     NaN   71546.00000                  NaN  
max                     NaN   99998.00000                  NaN

## Exploring some columns where there might be outlliers, like price and kilometer columns

### exploring and cleaning price column

In [12]:
data['price'].value_counts().sort_index()

0             10778
1              1189
2                12
3                 8
4                 1
              ...  
32545461          1
74185296          1
99000000          1
99999999         15
2147483647        1
Name: price, Length: 5597, dtype: int64

There are 10778 cars in our system that cost 0 dollars and there are also cars that cost more than 3 million dollars.

We can only allow cars that their vallue is between 100  and  3mil dollars

In [13]:
(~data['price'].between(100,3000000)).sum()/ data.shape[0]

0.035997286880127474

We also see that those extreme prices are only 3,6% of our overall data. So its fine to drop them.

In [14]:
#limitting the price between 100 and 3 mil dollars
data=data[data['price'].between(100,3000000)]

In [15]:
data['price'].value_counts().sort_index()

100        1032
101           4
105           3
106           1
108           1
           ... 
1250000       2
1300000       1
1600000       2
2795000       1
2995000       1
Name: price, Length: 5526, dtype: int64

### exploring kilometer column

In [16]:
#checking kilometer col
data['kilometer'].value_counts().sort_index()

5000        5502
10000       1828
20000       5494
30000       5916
40000       6312
50000       7520
60000       8581
70000       9646
80000      10873
90000      12326
100000     15371
125000     37232
150000    231553
Name: kilometer, dtype: int64

numbers look pretty normal in the kilometer column (min value= 5000km, max value=150000km) so we don't need to change anything

# Exploring the Dates

In [17]:
data[['year_of_registration','month_of_registration']].sort_values('year_of_registration')

year_of_registration  month_of_registration
159148                  1000                      0
102547                  1000                      0
142497                  1000                      0
58295                   1000                      0
74945                   1000                      0
...                      ...                    ...
30825                   9999                      0
238547                  9999                      0
30334                   9999                      0
315925                  9999                      0
210300                  9999                      0

[358154 rows x 2 columns]

In [18]:
data['month_of_registration'].value_counts()

3     35487
6     32378
0     31928
4     30222
5     29872
7     28313
10    26776
11    24993
12    24759
9     24551
1     23831
8     23189
2     21855
Name: month_of_registration, dtype: int64

In [19]:
data['year_of_registration'].value_counts().sort_index()

1000    23
1001     1
1039     1
1111     2
1234     4
        ..
8500     1
8888     2
9000     4
9450     1
9999    14
Name: year_of_registration, Length: 143, dtype: int64

So as we can see there are cars that are registered at month 0 which doesn't exist .

Also there are cars that are registered before 1900 ( aprox starting date when cars were being manufactured) and cars that are registered after 2017 which cannot be done ofc because the data in from 2016

In [20]:
(~data['month_of_registration'].between(1,12)).sum()/data.shape[0]

0.08914600981700609

In [21]:
(~data['year_of_registration'].between(1990,2016)).sum()/data.shape[0]

0.06559468831843285

The problematic month registration is covering 8,9% of our data

and the problematic year registration is covering 6,6 % of our data

We are going to keep only the car registration that are between 1990 and 2016 and then see if this also solves the month registration problem. Otherwise we are going to find a solution for it

In [22]:
data=data[data['year_of_registration'].between(1990,2016)]

In [23]:
(~data['year_of_registration'].between(1990,2016)).sum()/data.shape[0]

0.0

In [24]:
(~data['month_of_registration'].between(1,12)).sum()/data.shape[0]

0.08064877592548877

Solving the year of registration problem didn't help much with the month registration problem.

The thing is that we can't drop an additional 8% of our data in order to solve the month registration problem. Instaid, what we are going to do is assign the most common month of registration to the rows that have the month 0. Changing the month is not something that is going to change the value of a car that much, unlike the year which is more important for someone to consider when if he wants to buy a car

In [25]:
data['month_of_registration'].value_counts().sort_values()

2     20665
8     21966
1     22319
9     23318
12    23484
11    23796
10    25509
7     26232
0     26990
5     28147
4     28397
6     30403
3     33435
Name: month_of_registration, dtype: int64

March is the most common month of registration, so we are going to assing the month 0 to March

In [26]:
data['month_of_registration']=data['month_of_registration'].replace(0,3)

In [27]:
data['month_of_registration'].value_counts().sort_values()

2     20665
8     21966
1     22319
9     23318
12    23484
11    23796
10    25509
7     26232
5     28147
4     28397
6     30403
3     60425
Name: month_of_registration, dtype: int64

In [28]:
#making sure it worked
(~data['month_of_registration'].between(1,12)).sum()/data.shape[0]

0.0

### exploring our data for missing values

In [29]:
data.isnull().sum()

date_crawled                 0
name                         0
price                        0
ab_test                      0
vehicle_type             18436
year_of_registration         0
gearbox                  13838
power_ps                     0
model                    14403
kilometer                    0
month_of_registration        0
fuel_type                22201
brand                        0
not_repaired_damage      56749
date_created                 0
postal_code                  0
last_seen                    0
dtype: int64

We still have columns with missing values. We are going to explore them and see how we can fix that issue

In [30]:
data['vehicle_type'].value_counts()

limousine     90164
kleinwagen    76736
kombi         65439
bus           29159
cabrio        21176
coupe         16746
suv           14021
andere         2784
Name: vehicle_type, dtype: int64

In [31]:
data['gearbox'].value_counts()

manuell      249300
automatik     71523
Name: gearbox, dtype: int64

In [32]:
data['fuel_type'].value_counts()

benzin     203641
diesel     102821
lpg          5005
cng           532
hybrid        264
andere         99
elektro        98
Name: fuel_type, dtype: int64

In [33]:
data['model'].value_counts()

golf               26621
andere             21601
3er                19008
polo               11821
corsa              11424
                   ...  
elefantino             7
kalina                 6
rangerover             5
200                    4
discovery_sport        1
Name: model, Length: 248, dtype: int64

In [34]:
data['not_repaired_damage'].value_counts()

nein    246178
ja       31734
Name: not_repaired_damage, dtype: int64

By looking at the columns, the way i decided to solve the problem with missing values for each one is:
- veh_type: drop the rows since it is not covering that much of the overall data 
- gearbox: changing null values to manuel since its the most common by far and it is almost certain that someone who decided to not put any info into the gearbox section is going to be selling a manuel gearbox car
- fuel-type: drop the rows since we can't be certain about what type of fuel to assign to the missing ones and it is a huge selling point which we can't really change our data to whatever we believe is more common
- model: It is maybe the first thing someone is looking when he/she wants to buy a car, so we can't change it. That's why we have to drop these columns too
- not-rep-dmg: there is no reason to leave this empty other than your car that you are selling has been damaged in the past. Thus we are going to change it to 'ja'

In [35]:
data['gearbox']=data[data['gearbox'].isnull()]='manuel'

In [36]:
data['not_repaired_damage']=data[data['not_repaired_damage'].isnull()]='ja'

In [37]:
data.isnull().sum()

date_crawled                 0
name                         0
price                        0
ab_test                      0
vehicle_type             18436
year_of_registration         0
gearbox                      0
power_ps                     0
model                    14403
kilometer                    0
month_of_registration        0
fuel_type                22201
brand                        0
not_repaired_damage          0
date_created                 0
postal_code                  0
last_seen                    0
dtype: int64

In [38]:
data.shape[0]

334661

In [39]:
data=data.dropna(axis=0)

In [40]:
data.isnull().sum()

date_crawled             0
name                     0
price                    0
ab_test                  0
vehicle_type             0
year_of_registration     0
gearbox                  0
power_ps                 0
model                    0
kilometer                0
month_of_registration    0
fuel_type                0
brand                    0
not_repaired_damage      0
date_created             0
postal_code              0
last_seen                0
dtype: int64

In [41]:
data.shape[0]

293735

we dropped about 10% of our data in order to clear the missing value cells but we had no other choice

Now we can begin exploring the data and create models and or make predictions about the car prices etc